crates/scope/src/builder.rs

```diff
@@ -1654,6 +1659,164 @@ impl ScopeBuilderStack {
     }
 }
 
+/// Builds `FunctionStencil` for all functions (both non-lazy and lazy).
+/// `FunctionStencil.script` is set to lazy function, with inner functions and
+/// closed over bindings populated in gcthings list.
+///
+/// TODO: For non-lazy function, gcthings list should be populated in the
+///       emitter pass, not here.
+#[derive(Debug)]
+pub struct FunctionStencilBuilder {
+    /// The map from function node to FunctionStencil.
+    ///
+    /// It's separated into `function_stencil_indices` and `functions`,
+    /// because `FunctionStencil` is referred from multiple places:
+    ///   * map from Function AST node (`function_stencil_indices`)
+    ///   * enclosing script/function, to list inner functions
+    function_stencil_indices: AssociatedData<FunctionStencilIndex>,
+    functions: FunctionStencilList,
+
+    /// The stack of functions that the current context is in.
+    /// The innermost one is the current function that inner function is going
+    /// to be stored into.
+    function_stack: Vec<FunctionStencilIndex>,
+}
+
+impl FunctionStencilBuilder {
+    fn new() -> Self {
+        Self {
+            function_stencil_indices: AssociatedData::new(),
+            functions: FunctionStencilList::new(),
+            function_stack: Vec::new(),
+        }
+    }
+
+    /// Enter a function.
+    /// This creates `FunctionStencil` for the function, and adds it to
+    /// enclosing function is exists.
+    fn enter<T>(
+        &mut self,
+        fun: &T,
+        kind: FunctionKind,
+        is_lambda: bool,
```

Comment:
yeah, it would be nice to add struct, but "shape" means [different thing](https://searchfox.org/mozilla-central/rev/559b25eb41c1cbffcb90a34e008b8288312fcd25/js/src/vm/Shape.h#40) in SpiderMonkey, so I'd like to avoid.
I'll look for something that doesn't conflict.
(also we cannot use [kind](https://searchfox.org/mozilla-central/rev/559b25eb41c1cbffcb90a34e008b8288312fcd25/js/src/vm/FunctionFlags.h#23), and [syntax](https://searchfox.org/mozilla-central/rev/559b25eb41c1cbffcb90a34e008b8288312fcd25/js/src/frontend/FunctionSyntaxKind.h#15) .... :P

Reply:
err, I think we can reuse Syntax, given it's specific to the current frontend in m-c side, and won't conflict with jsparagus.